spanner/spannertest/db_query.go

```diff
@@ -744,190 +765,123 @@ func (ji *joinIter) prepUsing(using []spansql.ID, lhsEC, rhsEC evalContext) erro
 	}
 	ji.ec.row = make(row, len(ji.ec.cols))
 
-	ji.cond = func(lhs, rhs row) (bool, error) {
-		for i, lhsi := range lhsUsing {
-			rhsi := rhsUsing[i]
-			if compareVals(lhs[lhsi], rhs[rhsi]) != 0 {
-				return false, nil
-			}
-			ji.ec.row[i] = lhs[lhsi]
-		}
+	primaryUsing, secondaryUsing := lhsUsing, rhsUsing
+	if flipped {
+		primaryUsing, secondaryUsing = secondaryUsing, primaryUsing
+	}
 
-		// The loop above copied the values from the common columns into ji.ec.row already;
-		// we just need to copy the remaining values.
-		j := len(lhsUsing)
+	orNil := func(r row, i int) interface{} {
+		if r == nil {
+			return nil
+		}
+		return r[i]
+	}
+	// populate writes the data to ji.ec.row in the correct positions.
+	populate := func(primary, secondary row) { // secondary may be nil
+		j := 0
+		for _, pi := range primaryUsing {
+			ji.ec.row[j] = primary[pi]
+			j++
+		}
+		lhs, rhs := primary, secondary
+		if flipped {
+			rhs, lhs = lhs, rhs
+		}
 		for _, i := range lhsNotUsing {
-			ji.ec.row[j] = lhs[i]
+			ji.ec.row[j] = orNil(lhs, i)
 			j++
 		}
 		for _, i := range rhsNotUsing {
-			ji.ec.row[j] = rhs[i]
+			ji.ec.row[j] = orNil(rhs, i)
 			j++
 		}
-
+	}
+	ji.cond = func(primary, secondary row) (bool, error) {
+		for i, pi := range primaryUsing {
+			si := secondaryUsing[i]
+			if compareVals(primary[pi], secondary[si]) != 0 {
+				return false, nil
+			}
+		}
+		populate(primary, secondary)
 		return true, nil
 	}
+	ji.zero = func(primary row) {
+		populate(primary, nil)
+	}
 	return nil
 }
 
 type joinIter struct {
```

Comment:
I suggest starting to review from the changes to the `joinIter` struct; the rest should hopefully flow directly from that.